File: day6/solution.jl
# Read the input data into a matrix of characters
const data = stack(eachline(stdin), dims=1)
const nrows = size(data, 1)
const ncols = size(data, 2)

# Find initial position and direction
const pos = findfirst(==('^'), data)
const dir = '^'

# Direction deltas
const delta = Dict('^' => (-1, 0), '>' => (0, 1), '<' => (0, -1), 'v' => (1, 0))

# Part I: Walk in one direction until out of bounds or hitting a wall
function walk1(data, pos, dir)
    r = pos[1]
    c = pos[2]
    dr, dc = delta[dir]
    seen_pos = Set()
    inbounds = true
    while inbounds
        push!(seen_pos, (r, c))
        newr = r + dr
        newc = c + dc
        if !(1 <= newr <= nrows && 1 <= newc <= ncols)
            break
        end
        if data[newr,newc] == '#'
            dc, dr = -dr, dc
            continue
        end
        # print((r, c))
        r, c = newr, newc
    end
    return length(seen_pos)
end

println("Part I: ", walk1(data, pos, dir))

# Part II: Walk in one direction and detect cycles
function walk2(pos, dir)
    r = pos[1]
    c = pos[2]
    dr, dc = delta[dir]
    seen_state = Set()
    inbounds = true
    while inbounds
        if (r, c, dr, dc) in seen_state
            return true
        end
        push!(seen_state, (r, c, dr, dc))
        newr = r + dr
        newc = c + dc
        if !(1 <= newr <= nrows && 1 <= newc <= ncols)
            break
        end
        if data[newr, newc] == '#'
            dc, dr = -dr, dc
            continue
        end
        r, c = newr, newc
    end
    return false
end

function part2(data, pos, dir)
    total = 0
    for r in 1:nrows
        for c in 1:ncols
            if data[r,c] == '#' || data[r,c] == '^'
                continue
            end
            data[r,c] = '#'
            if walk2(pos, dir)
                total += 1
            end
            data[r,c] = '.'
        end
    end
    return total
end

println("Part II: ", part2(data, pos, dir))

File: day6/solution.py
data = [[c for c in l] for l in open(0).read().splitlines()]
nrows = len(data)
ncols = len(data[0])

def find_element(matrix, el):
    for row_idx, row in enumerate(matrix):
        if el in row:
            col_idx = row.index(el)
            return row_idx, col_idx
    return None

pos = find_element(data, '^')
dir = '^'

delta = {'^': (-1,0), '>': (0,1), '<': (0,-1), 'v': (1,0)}
# newdir = {'^': '>', '>': 'v', 'v': '<', '<': '^'}

def walk1(pos, dir):
    r, c = pos
    dr, dc = delta[dir]
    seen_pos = set()
    inbounds = True
    while inbounds:
        seen_pos.add((r, c))
        newr = r + dr
        newc = c + dc
        if not (0 <= newr < nrows and 0 <= newc < ncols):
            break
        if data[newr][newc] == '#':
            dc, dr = -dr, dc
            continue
        r = newr
        c = newc
    return len(seen_pos)

print("Part I:", walk1(pos, dir))



# PART II

def walk2(pos, dir):
    r, c = pos
    dr, dc = delta[dir]
    seen_state = set()
    inbounds = True
    while inbounds:
        if (r, c, dr, dc) in seen_state:
            return True
        seen_state.add((r, c, dr, dc))
        newr = r + dr
        newc = c + dc
        if not (0 <= newr < nrows and 0 <= newc < ncols):
            break
        if data[newr][newc] == '#':
            dc, dr = -dr, dc
            continue
        r = newr
        c = newc
    return False

# very slow but works....
total = 0
for r in range(nrows):
    for c in range(ncols):
        if data[r][c] == '#' or data[r][c] == '^':
            continue
        data[r][c] = '#'
        if walk2(pos, dir):
            total += 1
        data[r][c] = '.'
print("Part II:", total)